networksimulator.py

```diff
@@ -242,6 +258,81 @@ def clear_network(self):
         self._cur_time = 0
         self._num_active_flows = 0
         self._event_counter = 0
+
+    def record_buffer_occupancy(self, link_id, buffer_occupancy):
+        """Records a buffer occupancy data point.
+
+        Args:
+            link_id (str): the link id of the link this point belongs to.
+            buffer_occupancy (float): buffer_occupancy of the link at cur_time.
+
+        """
+        self.data_metrics.update_buffer_occupancy(link_id, buffer_occupancy, self.cur_time)
+
+    def record_packet_loss(self, link_id, num_packets_lost):
+        """Records a packet loss data point.
+
+        Args:
+            link_id (str): the link id of the link this point belongs to.
+            num_packets_lost (int): number of packets lost by the link at cur_time.
+
+        """
+        self.data_metrics.update_packet_loss(link_id, num_packets_lost, self.cur_time)
+
+    def record_link_rate(self, link_id, amt_sent):
```

Comment:
confusing a bit because it's called rate but the parameter is a number of bits and you don't seem to doing any division for bits / time